improv_ble_client/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+        """Subscribe to state updates."""
+        _LOGGER.debug("%s: subscribe_state_updates", self.name)
+
+        async def _subscribe_state_updates() -> Callable[[], None]:
+            return self._notification_handlers.subscribe_state(state_callback)
+
+        return await self._execute(_subscribe_state_updates)
+
+    @retry_bluetooth_connection_error(DEFAULT_ATTEMPTS)  # type: ignore[misc]
+    async def _execute(self, procedure: Callable[[], Coroutine[Any, Any, _T]]) -> _T:
+        """Execute a procedure."""
+        if self._procedure_lock.locked():
+            _LOGGER.debug(
+                "%s: Procedure already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        async with self._procedure_lock:
+            try:
+                await self._ensure_connected()
+                return await procedure()
+            except asyncio.CancelledError as err:
+                if self._disconnect_reason is None:
+                    raise ImprovError from err
+                if self._disconnect_reason == DisconnectReason.TIMEOUT:
+                    raise Timeout from err
+                if self._disconnect_reason == DisconnectReason.UNEXPECTED:
+                    raise UnexpectedDisconnect from err
+                raise Disconnected(self._disconnect_reason) from err
+            except ImprovError:
+                self._disconnect(DisconnectReason.ERROR)
+                raise
+
+    async def _ensure_connected(self) -> None:
+        """Ensure connection to device is established."""
+        if self._connect_lock.locked():
+            _LOGGER.debug(
+                "%s: Connection already in progress, waiting for it to complete; "
+                "RSSI: %s",
+                self.name,
+                self.rssi,
+            )
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        async with self._connect_lock:
+            # Check again while holding the lock
+            if self._client and self._client.is_connected:
+                self._reset_disconnect_timer()
+                return
+            _LOGGER.debug("%s: Connecting; RSSI: %s", self.name, self.rssi)
+            client = await establish_connection(
+                BleakClientWithServiceCache,
+                self._ble_device,
+                self.name,
+                self._disconnected,
+                use_services_cache=False,  # True
+                ble_device_callback=lambda: self._ble_device,
+            )
+            _LOGGER.debug("%s: Connected; RSSI: %s", self.name, self.rssi)
+
+            self._client = client
+            self._disconnect_reason = None
+            self._reset_disconnect_timer()
+
+            _LOGGER.debug(
+                "%s: Subscribe to notifications; RSSI: %s", self.name, self.rssi
+            )
+            await client.start_notify(
```

Comment:
Note that older esphome proxy devices are limited to 3 active notify handles at a time

Reply:
Is that a global limit? What's the limit for newer devices?

Reply:
> Is that a global limit? 
Yes, I believe so but the actual ESP-IDF code was not documented in english so I'm not 100% on that

> What's the limit for newer devices?

Its 7 or 9 for newer devices.

Reply:
OK. As discussed on Discord we'll not do anything about this right now, but it will be kept in mind in case onboarding is failing because of it.
Possible mitigations would be to poll some or all of the characteristics instead of subscribing.